src/variant_detection/variant_output.cpp

```diff
@@ -1,117 +1,194 @@
-                                tmp.set_alt("<INS>");
-                                tmp.add_info("SVTYPE", "INS");
-                                // Increment position by 1 because VCF is 1-based
-                                tmp.set_pos(mate1.position + 1);
-                                tmp.add_info("SVLEN", std::to_string(sv_length));
-                                // Increment end by 1 because VCF is 1-based
-                                tmp.add_info("END", std::to_string(mate1.position + 1));
-                                tmp.print(out_stream);
-                            }
-                        }
+                        record.alt().push_back("<DEL>");
+                        // Increment end by 1 because VCF is 1-based
+                        // Decrement end by 1 because deletion ends one base before mate2 begins
+                        record.info().push_back({.id = "END", .value = mate2.position});
+                        record.info().push_back({.id = "SVLEN", .value = sv_length});
+                        record.info().push_back({.id = "SVTYPE", .value = "DEL"});
+                        found_SV = true;
+                    }
+                    // Insertion (sv_length is positive)
+                    // for a small deletion inside of an insertion, the distance is a small negative value
+                    else if (sv_length > 0 &&
+                             distance <= 0 &&
+                             std::abs(distance) <= args.max_tol_deleted_length)
+                    {
+                        record.alt().push_back("<INS>");
+                        // Increment end by 1 because VCF is 1-based
+                        record.info().push_back({.id = "END", .value = mate1.position + 1});
+                        record.info().push_back({.id = "SVLEN", .value = sv_length});
+                        record.info().push_back({.id = "SVTYPE", .value = "INS"});
+                        found_SV = true;
                     }
                 }
             }
         }
     }
 }
 
-//!\overload
 void find_and_output_variants(std::map<std::string, int32_t> & references_lengths,
                               std::vector<Cluster> const & clusters,
                               cmd_arguments const & args,
                               std::filesystem::path const & output_file_path)
 {
+    bio::var_io::header hdr{};
+    write_header(references_lengths, args.vcf_sample_name, hdr);
+
+    bool found_SV = false;
+
     if (output_file_path.empty())
     {
-        find_and_output_variants(references_lengths, clusters, args, std::cout);
+        std::ostringstream stream{};
+        bio::var_io::writer writer{stream, bio::vcf{}};
+        writer.set_header(hdr);
+        for (size_t i = 0; i < clusters.size(); ++i)
+        {
+            // ignore low quality SVs
+            if (clusters[i].get_cluster_size() >= args.min_qual)
+            {
+                bio::var_io::default_record<> record{};
```

Comment:
You might still be able to move this record out of the loop completely. But only if you always overwrite or clear the fields.